File: 94_BinaryTreeInOrderTraversal/Solution.cpp
#include "../IOLib.hpp"

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> result;
        traverse(root, result);
        return result;
    }
    void traverse(TreeNode* root, vector<int>& nums){
        if(root == NULL) return;
        if(root->left != NULL) traverse(root->left, nums);
        nums.push_back(root->val);
        if(root->right != NULL) traverse(root->right, nums);
    }
};

int main(){
	TreeNode* root = readBinaryTree();
	Solution solution;
	vector<int> nums = solution.inorderTraversal(root);
	printVector(nums);
	return 0;
}

File: 94_BinaryTreeInOrderTraversal/Solution.java
import java.util.List;
import java.util.ArrayList;
import java.util.Stack;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public static List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> list = new ArrayList<Integer>();
        if(root == null) return list;
        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode ptr = root;
        while(ptr != null || !stack.empty()){
            while(ptr != null){
                stack.push(ptr);
                ptr = ptr.left;
            }
            ptr = stack.pop();
            list.add(ptr.val);
            ptr = ptr.right;
        }
        return list;
    }
}